coins/round.go

```diff
@@ -5,18 +5,107 @@ package coins
 
 import (
 	"github.com/cockroachdb/apd/v3"
+	"github.com/ethereum/go-ethereum/common/math"
 )
 
-func roundToDecimalPlace(result *apd.Decimal, n *apd.Decimal, decimalPlace uint8) error {
-	result.Set(n) // already optimizes result == n
+func roundToDecimalPlace(n *apd.Decimal, decimalPlace uint8) (*apd.Decimal, error) {
```

Comment:
The previous version had a defect in that it assumed that the `result` was not the same memory address as `n` (the number being rounded). We don't care about optimizing the number of apd.Decimal's that are allocated, so I just simplified things and changed the API to always return a freshly allocated value.